File: transcribe-service/app/api/google-drive.js
import fs from "fs";
import { google } from "googleapis";

import { GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS_FILE_PATH } from "../constants.js";

if (!fs.existsSync(GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS_FILE_PATH)) {
  fs.writeFileSync(
    GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS_FILE_PATH,
    process.env.GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS
  );
}

const auth = new google.auth.GoogleAuth({
  keyFile: GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS_FILE_PATH,
  scopes: ["https://www.googleapis.com/auth/drive"],
});

const drive = google.drive({
  version: "v3",
  auth,
});

export async function uploadToGoogleDrive({ data, filename }) {
  fs.writeFileSync(filename, data, { flag: "w" });

  let response;
  try {
    response = await drive.files.create({
      requestBody: {
        name: filename,
        fields: "id",
        parents: [process.env.GOOGLE_DRIVE_SRT_FOLDER_ID],
      },
      media: {
        mimeType: "text/plain",
        body: fs.createReadStream(filename),
      },
    });
  } catch (error) {
    console.error("Error uploading file to Google Drive:", error);
    throw error;
  } finally {
    fs.unlink(filename, (error) => {
      if (error) {
        console.error("Error deleting file:", error);
      }
    });
  }
  return { fileId: response?.data?.id };
}

export async function getFileInfo({ fileId }) {
  let response;
  try {
    response = await drive.files.get({
      fileId,
      fields: "name,webViewLink",
    });
  } catch (error) {
    console.error("Error getting file info from Google Drive:", error);
    throw error;
  }
  return {
    name: response?.data?.name,
    webViewLink: response?.data?.webViewLink,
  };
}

export async function setPermission({ fileId }) {
  try {
    await drive.permissions.create({
      fileId,
      requestBody: {
        role: "reader",
        type: "anyone",
      },
    });
  } catch (error) {
    console.log("Failed to set permission on file", fileId, error.message);
  }
}

File: translate/src/api/google_drive.py
import os
from google.oauth2 import service_account
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload

from ..types import FileUploadRequest

SCOPES = ["https://www.googleapis.com/auth/drive"]
SERVICE_ACCOUNT_FILE = './service-account.secret.json'

if not os.path.exists(SERVICE_ACCOUNT_FILE):
  with open(SERVICE_ACCOUNT_FILE, 'w') as f:
    credentials = os.getenv('GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS')
    f.write(os.getenv('GOOGLE_DRIVE_SERVICE_ACCOUNT_CREDENTIALS') or "")

credentials = service_account.Credentials.from_service_account_file(
  SERVICE_ACCOUNT_FILE,
  scopes=SCOPES
)

drive_service = build('drive', 'v3', credentials=credentials)

def upload_to_google_drive(params: FileUploadRequest) -> dict:
  
  text, file_name, properties = params.text, params.file_name, params.properties

  with open(file_name, 'w') as f:
    f.write(text)

  try:
    file_metadata = {
      'name': file_name,
      'parents': [os.getenv('GOOGLE_DRIVE_SRT_FOLDER_ID')],
      'properties': properties
    }
    media = MediaFileUpload(file_name, mimetype='text/plain')
    file = drive_service.files().create(
      body=file_metadata,
      media_body=media,
      fields='id'
    ).execute()
    print(f"Uploaded file to Google Drive: {file.get('id')}")
  except Exception as error:
    print(f"Error uploading file to Google Drive: {error}")
    raise error
  finally:
    try:
      os.remove(file_name)
    except Exception as error:
      print(f"Error deleting file: {error}")

  return {'file_id': file.get('id')}

def get_file_info(file_id: str) -> dict:
  try:
    file = drive_service.files().get(
      fileId=file_id,
      fields='name,webViewLink,properties'
    ).execute()
  except Exception as error:
    print(f"Error getting file info from Google Drive: {error}")
    raise error

  return {
    'name': file.get('name'),
    'webViewLink': file.get('webViewLink'),
    'properties': file.get('properties')
  }
